plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2885,36 +2892,52 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
+        IcebergTableHandle cacheKey = new IcebergTableHandle(
+                originalHandle.getCatalog(),
+                originalHandle.getSchemaName(),
+                originalHandle.getTableName(),
+                originalHandle.getTableType(),
+                originalHandle.getSnapshotId(),
+                originalHandle.getTableSchemaJson(),
+                originalHandle.getPartitionSpecJson(),
+                originalHandle.getFormatVersion(),
+                originalHandle.getUnenforcedPredicate(),
+                originalHandle.getEnforcedPredicate(),
+                OptionalLong.empty(), // limit is currently not included in stats and is not enforced by the connector
+                ImmutableSet.of(), // projectedColumns are used to request statistics only for the required columns, but are not part of cache key
+                originalHandle.getNameMappingJson(),
+                originalHandle.getTableLocation(),
+                originalHandle.getStorageProperties(),
+                false, // recordScannedFiles does not affect stats
+                originalHandle.getMaxScannedFileSize(),
+                ImmutableSet.of(), // constraintColumns do not affect stats
+                Optional.empty()); // forAnalyze does not affect stats
+        return getIncrementally(
+                tableStatisticsCache,
+                cacheKey,
+                currentStatistics -> currentStatistics.getColumnStatistics().keySet().containsAll(originalHandle.getProjectedColumns()),
+                () -> {
+                    Table icebergTable = catalog.loadTable(session, originalHandle.getSchemaTableName());
                     return TableStatisticsReader.getTableStatistics(
                             typeManager,
                             session,
-                            handle,
+                            originalHandle,
+                            originalHandle.getProjectedColumns(),
                             icebergTable,
                             fileSystemFactory.create(session.getIdentity(), icebergTable.io().properties()));
-                });
+                },
+                currentStatistics -> {
+                    Sets.SetView<IcebergColumnHandle> missingColumns = difference(originalHandle.getProjectedColumns(), currentStatistics.getColumnStatistics().keySet());
+                    Table icebergTable = catalog.loadTable(session, originalHandle.getSchemaTableName());
+                    return TableStatisticsReader.getTableStatistics(
+                            typeManager,
+                            session,
+                            originalHandle,
+                            missingColumns,
+                            icebergTable,
+                            fileSystemFactory.create(session.getIdentity(), icebergTable.io().properties()));
+                },
+                IcebergMetadata::mergeColumnStatistics);
     }
```

Comment:
while trying to understand the logic of `getIncrementally` method , i am reading the method in split mode  to see what the loader, merge, incrementalLoader are doing.

optional: Am advocating for a non-generic version of `getIncrementally`


```
private static TableStatistics getIncrementally(
            Map<IcebergTableHandle, AtomicReference<TableStatistics>> cache,
            IcebergTableHandle cacheKey,
            Predicate<TableStatistics> isSufficient,
            Supplier<TableStatistics> loader,
            Function<TableStatistics, TableStatistics> incrementalLoader)
    {
        AtomicReference<TableStatistics> valueHolder = cache.computeIfAbsent(cacheKey, _ -> new AtomicReference<>());
        TableStatistics oldValue = valueHolder.get();
        if (oldValue != null && isSufficient.test(oldValue)) {
            return oldValue;
        }

        TableStatistics newValue = oldValue == null ? loader.get() : incrementalLoader.apply(oldValue);

        verifyNotNull(newValue, "loader returned null for %s", cacheKey);

        TableStatistics merged = mergeColumnStatistics(oldValue, newValue);
        if (!valueHolder.compareAndSet(oldValue, merged)) {
            // if the value changed in the valueHolder, we only add newly loaded value to be sure we have up-to-date value
            valueHolder.accumulateAndGet(newValue, IcebergMetadata::mergeColumnStatistics);
        }
        return merged;
    }
```

Do note that the `loader` and `incrementalLoader` differ from each other only in  one specific line - the line providing the projected columns.
We could simply do `loader.apply(oldValue)` and encapsulate the initial/incremental logic into it.

Reply:
simplified it